spec/core/ics-033-multi-hop/README.md

```diff
@@ -0,0 +1,541 @@
+7. Repeat step 5, proving `ConsensusState[i]`, and `Conn[i,i+1]` where `i` is the proof index starting with the consensus state on `Chain[N-2]`. `ConsensusState[N-1]` is already known on `Chain[N]`. Note that chains are indexed from sender to receiver and proofs are indexed in the opposite direction to match the connectionHops ordering.
+   - Verify ParseClientID(ConsensusProofs[i].Key) == ConnectionEnd.ClientID
+   - ConsensusProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
+   - ConnectionProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConnectionProofs[i].Key, ConnectionProofs[i].Value)
+   - ConsensusState = ConsensusProofs[i].Value
+   - i++
+8. Finally, prove the expected channel or packet commitment in `ConsensusState[N-2]` (sending chain consensus state) on `Chain[1]`
+
+For more details see [ICS4](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics).
+
+### Multi-hop Proof Verification Pseudo Code
+
+Pseudocode proof generation for a channel between `N` chains `C[0] --> C[i] --> C[N]`
+
+```go
+// VerifyMultihopMembership verifies a multihop membership proof.
+// Inputs: consensusState - The consensusState for chain[N-1], which is known on the destination chain (chain[N]).
+//         connectionHops - The expected connectionHops for the channel from the source chain to the destination chain.
+//         proof          - The serialized multihop proof data.
+//         prefix         - Merkleprefix to be combined with key to generate Merklepath for the key/value proof verification.
+//         key            - The key to prove in the indexed consensus state.
+//         value          - The value to prove in the indexed consensus state.
+func VerifyMultihopMembership(
+    consensusState exported.ConsensusState,
+    connectionHops []string,
+    proof []byte,
+    prefix exported.Prefix,
+    key string,
+    value []byte,
+) {
+    // deserialize proof bytes into multihop proofs
+    proofs := abortTransactionUnless(Unmarshal(proof))
+    abortTransactionUnless(len(proofs.ConsensusProofs) >= 1)
+    abortTransactionUnless(len(proofs.ConnectionProofs) == len(proofs.ConsensusProofs))
+
+    // verify connection hop ordering and connections are in OPEN state
+    abortTransactionUnless(VerifyConnectionHops(proofs.ConnectionProofs, connectionHops))
+
+    // verify intermediate consensus and connection states from destination --> source
+    abortTransactionUnless(VerifyConsensusAndConnectionStates(consensusState, proofs.ConsensusProofs, proofs.ConnectionProofs))
+
+    // verify a key/value proof on source chain's consensus state.
+    abortTransactionUnless(VerifyKeyMembership(proofs, prefix, key, value))
+}
+
+// VerifyMultihopNonMembership verifies a multihop non-membership proof.
+// Inputs: consensusState - The consensusState for chain[N-1], which is known on the destination chain (chain[N]).
```

Comment:
This would imply a reversal of the definition of chain ordering (currently 0 == Sender, N == Receiver). Proof ordering was changed per a previous feedback so that they match the channel hops, but I'm not sure we want to change the chain ordering (side effects would be to rewrite all comments on proof logic and recreate the proof diagrams).

👍🏼 on "destination" --> "receiving"